pennylane/devices/default_qubit.py

```diff
@@ -336,6 +337,41 @@ def _apply_cnot(self, state, axes, **kwargs):
         state_x = self._apply_x(state[sl_1], axes=target_axes)
         return self._stack([state[sl_0], state_x], axis=axes[0])
 
+    def _apply_toffoli(self, state, axes, **kwargs):
+        """Applies a Toffoli gate by slicing along the first axis specified in ``axes``, slicing
+        each of the resulting sub-arrays along the same axis, and then applying an X transformation
+        along the second axis of the fourth sub-sub-array.
+
+        By slicing along the first axis twice, we are able to select all of the amplitudes with a
+        corresponding :math:`|11\rangle` for the two control qubits. This means we then just need
+        to apply a :class:`~.PauliX` (NOT) gate to the result.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        sl_a0 = _get_slice(0, axes[0], self.num_wires)
+        sl_a1 = _get_slice(1, axes[0], self.num_wires)
+
+        # If axes[2] is larger than axes[0], then we need to shift the target axis down by
+        # two, otherwise we can leave as-is.
+        if axes[2] > axes[0]:
+            sl_b0 = _get_slice(0, axes[0], self.num_wires - 1)
+            sl_b1 = _get_slice(1, axes[0], self.num_wires - 1)
+            target_axes = [axes[2]-2]
+        else:
+            sl_b0 = _get_slice(0, axes[1], self.num_wires - 1)
+            sl_b1 = _get_slice(1, axes[1], self.num_wires - 1)
+            target_axes = [axes[2]]
+
+        # state[sl_a1][sl_b1] gives us all of the amplitudes with a |11> for the two control qubits.
+        state_x = self._apply_x(state[sl_a1][sl_b1], axes=target_axes)
+        state_out = self._stack([state[sl_a0][sl_b0], state[sl_a0][sl_b1], state[sl_a1][sl_b0], state_x])
+        return state_out.reshape((2, 2, 2))
```

Comment:
Care should be taken, as this reshaping now only assumes a 3 qubit system (see also the log for the [test checking the application of Toffoli](https://github.com/PennyLaneAI/pennylane/pull/1249/checks?check_run_id=2420120769#step:10:53)).

 However, we could have more than 3 qubits too. One solution is to make the reshaping depend on the number of qubits (stored as `self.num_wires`)
```suggestion
        return state_out.reshape([2] * self.num_wires)
```